README.md
# MMM-uptimerobot [![GitHub license](https://img.shields.io/badge/license-MIT-blue.svg)](https://github.com/mrVragec/MMM-uptimerobot/raw/master/LICENSE) [![Build Status](https://api.travis-ci.org/mrVragec/MMM-uptimerobot.svg?branch=master)](https://travis-ci.org/mrVragec/MMM-uptimerobot) [![Known Vulnerabiliries](https://snyk.io/test/github/mrvragec/mmm-uptimerobot/badge.svg)](https://snyk.io/test/github/mrvragec/mmm-uptimerobot) [![Code Climate](https://codeclimate.com/github/mrVragec/MMM-uptimerobot/badges/gpa.svg)](https://codeclimate.com/github/mrVragec/MMM-uptimerobot)

MagicMirror² to get uptime data from http://uptimerobot.com API. 

## Examples
![](.github/text.png) ![](.github/text_color.png) ![](.github/icons.png) ![](.github/icons_color.png)


## Instalation
1. Clone this repository in your MagicMirror installation under modules.
2. Rename plugin foler from 'MMM-uptimerobot' to 'uptimerobot'
3. Add configuration to config.js

## Using the module

To use this module, add the following configuration block to the modules array in the `config/config.js` file:
```js
var config = {
    modules: [
        {
            module: 'uptimerobot',
            config: {
                // See below for configurable options
            }
        }
    ]
}
```
## Requirements

No special requirements or dependencies. 

## Configuration options

| Option           | Description
|----------------- |-----------
| `api_key`        | *Required* API Key from https://uptimerobot.com. API Key is available under 'My Settings' in 'API Setting' section (click on Show/hide it and if not available generate new one).
| `useIcons`       | *Optional* Flag to use icons (true) or text (false) values for status of monitor. <br><br>**Type:** `boolean` <br>Default: false - text values.
| `useColors`      | *Optional* Flag to use colors (true) for status of monitor<br><br>**Type:** `boolean` <br>Default: false - without colors.
| `maximumEntries` | *Optional* Number of items (monitors) to shown. <br><br>**Type:** `int` <br>Default: 10 items.
| `statuses`       | *Optional* Filter out monitor by status (0 - paused; 1 - not checked yet; 2 - up; 8 - seems down; 9 - down<br><br>**Type:** `String` <br>Default: 0-1-2-8-9 (show monitors with all statuses).


CHANGELOG.md
## [1.2.0] - 2018-02-25

- add French, German and Spanish translation
- fix naming confusion with MMM name prefix
- fix register name

## [1.0.0] - 2017-04-27

First public release


node_helper.js
/* Magic Mirror
 * Node Helper: uptimerobot
 *
 * By Simon Crnko
 * MIT Licensed.
 */

var NodeHelper = require("node_helper");
var request = require('request');

module.exports = NodeHelper.create({

    start: function () {
        this.config = null;
    },

    // Override socketNotificationReceived method.

    /* socketNotificationReceived(notification, payload)
     * This method is called when a socket notification arrives.
     *
     * argument notification string - The identifier of the noitication.
     * argument payload mixed - The payload of the notification.
     */
    socketNotificationReceived: function (notification, payload) {
        if (notification === "uptimerobot-getData") {
            this.config = payload;
            this.getData();
        }
    },

    /*
     * getData
     * function example return data and show it in the module wrapper
     * get a URL request
     *
     */
    getData: function (config) {
        var self = this;

        var urlApi = "https://api.uptimerobot.com/v2/getMonitors";
        var retry = true;

        var options = {
            method: 'POST',
            url: 'https://api.uptimerobot.com/v2/getMonitors',
            headers: {
                'cache-control': 'no-cache',
                'content-type': 'application/x-www-form-urlencoded'
            },
            form: {
                api_key: this.config.api_key,
                format: 'json',
                logs: '0',
                // Limit the max number of records to return for the response.
                limit: this.config.maximumEntries,
                statuses: this.config.statuses
            }
        };
        request(options, function (error, response, body) {
            if (error) {
                retry = false;
                throw new Error(error);
            }
            self.sendSocketNotification("uptimerobot-processData", JSON.parse(body));
            if (retry) {
                self.scheduleUpdate((self.loaded) ? -1 : self.config.retryDelay);
            }
        });
    },

    /* scheduleUpdate()
     * Schedule next update.
     *
     * argument delay number - Milliseconds before next update.
     *  If empty, this.config.updateInterval is used.
     */
    scheduleUpdate: function (delay) {
        var nextLoad = this.config.updateInterval;
        if (typeof delay !== "undefined" && delay >= 0) {
            nextLoad = delay;
        }
        nextLoad = nextLoad;
        var self = this;
        setTimeout(function () {
            self.getData();
        }, nextLoad);
    }

});


uptimerobot.js
/* Magic Mirror
 * Module: uptimerobot
 *
 * By Simon Crnko
 * MIT Licensed.
 */

Module.register("uptimerobot", {
  defaults: {
    updateInterval: 60000,
    retryDelay: 5000,
    useIcons: false,
    useColors: false,
    maximumEntries: 10,
    statuses: "0-1-2-8-9"
  },

  requiresVersion: "2.1.0", // Required version of MagicMirror

  start: function () {
    var self = this;
    var dataRequest = null;

    //Flag for check if module is loaded
    this.loaded = false;
    if (this.config.api_key !== undefined) {
      // Schedule update timer.
      this.getData();
      setInterval(function () {
        self.updateDom();
      }, this.config.updateInterval);
    }
  },

  getStyles: function () {
    return ['uptimerobot.css', 'modules/uptimerobot/css/font-awesome.css'];
  },

  getData: function () {
      this.sendSocketNotification("uptimerobot-getData", this.config);
  },

  createWrapper: function(textToTranslate) {
    var wrapperDataNotification = document.createElement("div");
    wrapperDataNotification.innerHTML = this.translate(textToTranslate);
    return wrapperDataNotification;
  },

  // getScripts: function () {
  //   return [];
  // },

  // Load translations files
  getTranslations: function () {
    //FIXME: This can be load a one file javascript definition
    return {
      en: "translations/en.json",
      es: "translations/es.json",
      de: "translations/de.json",
      fr: "translations/fr.json"
    };
  },

  processData: function (data) {
    this.dataRequest = data;
    if (this.loaded === false) {
      this.updateDom(this.config.animationSpeed);
    }
    this.loaded = true;
  },

  // socketNotificationReceived from helper
  socketNotificationReceived: function (notification, payload) {
    if (notification === "uptimerobot-processData") {
      this.processData(payload);
        this.updateDom();
    }
  },

  setStatus: function (statusObject, statusForTranslate, statusClass, status) {
    if (!this.config.useIcons) {
      statusObject.innerHTML = this.translate(statusForTranslate);
    } else {
      statusObject.className = statusClass;
    }

    statusObject.className += " " + status;

    if (this.config.useColors) {
      statusObject.className += " colored";
    }
  },

  getStatusTest: function (statusValue) {
    var status = document.createElement('td');
    switch (statusValue) {
      case 0:
        this.setStatus(status, "PAUSED", "fa fa-pause-circle-o", "paused")
        break;
      case 1:
        this.setStatus(status, "NOTCHECKEDYET", "fa fa-retweet", "not-checked-yet");
        break;
      case 2:
        this.setStatus(status, "ONLINE", "fa fa-arrow-circle-up", "online");
        break;
      case 8:
        this.setStatus(status, "SEEMSDOWN", "fa fa-chevron-circle-down", "seems-down");
        break;
      case 9:
        this.setStatus(status, "DOWN", "fa fa-arrow-circle-down", "offline");
        break;
      default:
        return "";
    }
    return status;
  },


  getDom: function () {
    var self = this;
    // create element wrapper for show into the module
    var wrapper = document.createElement("div");

    // If this.dataRequest is not empty
    if (self.dataRequest) {
      var innerTable = document.createElement("table");
      innerTable.className = "small";

      self.dataRequest.monitors.forEach(function (element) {
        // create new row for each item in array
        var tableLine = document.createElement("tr");

        // create a cell in a row for name of server
        var lineCell = document.createElement("td");

        lineCell.className = 'friendlyName';
        lineCell.innerHTML = element.friendly_name;
        tableLine.appendChild(lineCell);

        // add status
        tableLine.appendChild(self.getStatusTest(element.status));
        innerTable.appendChild(tableLine);
      });

      wrapper.appendChild(innerTable);
    } else if (self.config.api_key === undefined) {
      // Missing API KEY
      //var wrapperDataNotification = document.createElement("div");
      // translations  + datanotification
      //wrapperDataNotification.innerHTML = self.translate("MISSING_API_KEY");
      wrapper.appendChild(self.createWrapper("MISSING_API_KEY"));
    } else {
      // Loading
      wrapper.className = "dimmed light small";
      wrapper.appendChild(self.createWrapper("LOADING"));
    }

    return wrapper;
  }
});


MMM-uptimerobot.js
/* Magic Mirror
 * Module: MMM-uptimerobot
 *
 * By Simon Crnko
 * MIT Licensed.
 */

Module.register("MMM-uptimerobot", {
  defaults: {
    updateInterval: 60000,
    retryDelay: 5000,
    useIcons: false,
    useColors: false,
    maximumEntries: 10,
    statuses: "0-1-2-8-9"
  },

  requiresVersion: "2.1.0", // Required version of MagicMirror

  start: function () {
    var self = this;
    var dataRequest = null;

    //Flag for check if module is loaded
    this.loaded = false;
    if (this.config.api_key !== undefined) {
      // Schedule update timer.
      this.getData();
      setInterval(function () {
        self.updateDom();
      }, this.config.updateInterval);
    }
  },

  getStyles: function () {
    return ['MMM-uptimerobot.css', 'modules/MMM-uptimerobot/css/font-awesome.css'];
  },

  getData: function () {
      this.sendSocketNotification("uptimerobot-getData", this.config);
  },

  createWrapper: function(textToTranslate) {
    var wrapperDataNotification = document.createElement("div");
    wrapperDataNotification.innerHTML = this.translate(textToTranslate);
    return wrapperDataNotification;
  },

  // getScripts: function () {
  //   return [];
  // },

  // Load translations files
  getTranslations: function () {
    //FIXME: This can be load a one file javascript definition
    return {
      en: "translations/en.json",
      es: "translations/es.json",
      de: "translations/de.json",
      fr: "translations/fr.json"
    };
  },

  processData: function (data) {
    this.dataRequest = data;
    if (this.loaded === false) {
      this.updateDom(this.config.animationSpeed);
    }
    this.loaded = true;
  },

  // socketNotificationReceived from helper
  socketNotificationReceived: function (notification, payload) {
    if (notification === "uptimerobot-processData") {
      this.processData(payload);
        this.updateDom();
    }
  },

  setStatus: function (statusObject, statusForTranslate, statusClass, status) {
    if (!this.config.useIcons) {
      statusObject.innerHTML = this.translate(statusForTranslate);
    } else {
      statusObject.className = statusClass;
    }

    statusObject.className += " " + status;

    if (this.config.useColors) {
      statusObject.className += " colored";
    }
  },

  getStatusTest: function (statusValue) {
    var status = document.createElement('td');
    switch (statusValue) {
      case 0:
        this.setStatus(status, "PAUSED", "fa fa-pause-circle-o", "paused")
        break;
      case 1:
        this.setStatus(status, "NOTCHECKEDYET", "fa fa-retweet", "not-checked-yet");
        break;
      case 2:
        this.setStatus(status, "ONLINE", "fa fa-arrow-circle-up", "online");
        break;
      case 8:
        this.setStatus(status, "SEEMSDOWN", "fa fa-chevron-circle-down", "seems-down");
        break;
      case 9:
        this.setStatus(status, "DOWN", "fa fa-arrow-circle-down", "offline");
        break;
      default:
        return "";
    }
    return status;
  },


  getDom: function () {
    var self = this;
    // create element wrapper for show into the module
    var wrapper = document.createElement("div");

    // If this.dataRequest is not empty
    if (self.dataRequest) {
      var innerTable = document.createElement("table");
      innerTable.className = "small";

      self.dataRequest.monitors.forEach(function (element) {
        // create new row for each item in array
        var tableLine = document.createElement("tr");

        // create a cell in a row for name of server
        var lineCell = document.createElement("td");

        lineCell.className = 'friendlyName';
        lineCell.innerHTML = element.friendly_name;
        tableLine.appendChild(lineCell);

        // add status
        tableLine.appendChild(self.getStatusTest(element.status));
        innerTable.appendChild(tableLine);
      });

      wrapper.appendChild(innerTable);
    } else if (self.config.api_key === undefined) {
      // Missing API KEY
      //var wrapperDataNotification = document.createElement("div");
      // translations  + datanotification
      //wrapperDataNotification.innerHTML = self.translate("MISSING_API_KEY");
      wrapper.appendChild(self.createWrapper("MISSING_API_KEY"));
    } else {
      // Loading
      wrapper.className = "dimmed light small";
      wrapper.appendChild(self.createWrapper("LOADING"));
    }

    return wrapper;
  }
});
